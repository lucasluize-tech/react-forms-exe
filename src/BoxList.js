import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from "uuid";

const BoxList = () => {
  const [boxes, setBoxes] = useState([]);

  const addBox = (box) => {
    let newBox = { ...box, id: uuid() };
    setBoxes((boxes) => [...boxes, newBox]);
  };

  const handleDelete = (event) => {
    const box = event.target.previousSibling;
    const boxRemoved = boxes.filter((b) => b.id !== box.id);
    setBoxes(boxRemoved);
  };

  return (
    <div
      className='BoxList'
      style={{
        display: "flex",
        flexDirection: "column",
        justifyContent: "center",
        alignItems: "center",
      }}>
      <NewBoxForm addBox={addBox} />

      {boxes.map((box) => (
        <Box
          key={box.id}
          n={box.id}
          width={box.width}
          height={box.height}
          color={box.color}
          handleDelete={handleDelete}
        />
      ))}
    </div>
  );
};

export default BoxList;
