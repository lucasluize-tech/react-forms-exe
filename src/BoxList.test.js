import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, height = "20", width = "20", color = "peachpuff") {
  const heightInput = boxList.getByLabelText("Height");
  const widthInput = boxList.getByLabelText("Width");
  const backgroundInput = boxList.getByLabelText("Color:");
  fireEvent.change(backgroundInput, { target: { value: color } });
  fireEvent.change(widthInput, { target: { value: width } });
  fireEvent.change(heightInput, { target: { value: height } });
  const button = boxList.getByText("Add Box!");
  fireEvent.click(button);
}

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function () {
  const boxList = render(<BoxList />);

  // no boxes yet
  expect(boxList.queryByText("X")).not.toBeInTheDocument();

  addBox(boxList);

  // expect to see a box
  const removeButton = boxList.getByText("Add Box!");
  expect(removeButton).toBeInTheDocument();
  // expect(removeButton.previousSibling).toHaveStyle(`
  //   width: 20px;
  //   height: 20px;
  //   background-color: peachpuff;
  // `);
});

it("can remove a box", function () {
  const boxList = render(<BoxList />);
  addBox(boxList);

  const removeButton = boxList.getByText("X");

  // click the remove button and the box should be gone
  fireEvent.click(removeButton);
  expect(removeButton).not.toBeInTheDocument();
});
