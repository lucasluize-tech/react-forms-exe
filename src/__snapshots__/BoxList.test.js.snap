// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
    style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
  >
    <form>
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value="0"
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value="0"
      />
      <label
        for="color"
      >
        Color:
      </label>
      <input
        id="color"
        name="color"
        type="color"
        value="black"
      />
      <button
        type="submit"
      >
        Add Box!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
